immutablecollections/_immutabledict.py

```diff
@@ -66,13 +69,46 @@ def immutabledict(
             f"Cannot create an immutabledict from {type(iterable)}, only {InstantiationTypes}"
         )
 
+    if forbid_duplicate_keys:
+        keys: List[KT]
+        # `for x in dict` grabs keys, but `for x in pairs` grabs pairs, so we must branch
+        if isinstance(iterable, Mapping):
+            # duplicate keys are possible if input is e.g. a multidict
+            keys = list(iterable.keys())
+        else:
+            # iterable must be a (key, value) pair iterable
+            iterable = list(iterable)  # in case iterable is consumed by iteration
+            keys = [key for key, value in iterable]
+        seen: Set[KT] = set()
+        duplicated: Set[KT] = set()
+        for key in keys:
```

Comment:
So, this is really super-slow compared to the call to `dict()` hidden inside `_RegularDictBackedImmutableDict` (which is implemented in C).  Instead, I'd suggest doing a two-part approach:

* if `forbid_duplicate_keys` is `True`, then `list`-ify the input `Iterable`.
* *after* we construct `ret`, then if `forbid_duplicate_keys` is `True`, compare then `len` of the `ret` to the `len` of the input iterable. If there is a mismatch, then do the work to construct the error message